interfaces/acados_template/acados_template/mpc_utils.py

```diff
@@ -28,37 +28,355 @@
+
+        if len(slack_cost) == 0:
+            cost = cost_without_slacks
+        else:
+            cost = cost_without_slacks + slack_cost
+
+        # evaluate sum
+        result = {
+            'cost': array_to_float(cost),
+            'cost_without_slacks': array_to_float(cost_without_slacks),
+            'cost_slacks': array_to_float(slack_cost),
+            'violation_soft_constraints': np.concatenate((lower_slack.full(), upper_slack.full())),
+            'violation_hard_constraints': violation_hard_constraints,
+        }
+        return result
+
+    def evaluate_ocp_cost(self, acados_ocp_iterate: AcadosOcpIterate):
+        """
+        Evaluates the cost of a whole OCP trajectory, as evaluated inside acados.
+        """
+        acados_ocp_iterate = deepcopy(acados_ocp_iterate)
+        acados_ocp_iterate.flatten()
+        x_traj = np.array(acados_ocp_iterate.x_traj)
+        u_traj = np.array(acados_ocp_iterate.u_traj)
+
+        N, nx = np.array(acados_ocp_iterate.x_traj).shape
+        cost = 0
+
+        # the cost on the first step is different in the OCP
+        result = self.evaluate(x_traj[0, :], u_traj[0, :], step=0)
+        cost += result['cost_without_slacks']
+
+        for i in range(1,N-1):
+            result = self.evaluate(x_traj[i,:],u_traj[i,:],step=i)
+            cost += result['cost']
+
+        cost_fun_args = [x_traj[N-1,:], self.__parameter_values[:,-1], self.__p_global_values]
+        cost += self.terminal_cost_fun(*cost_fun_args).full()
+
+        lower_violation_e, upper_violation_e, lower_slack_e, upper_slack_e = (
+            self.constraint_function_e(x_traj[N-1,:],
+                                     self.__parameter_values[:,-1],
+                                     self.__p_global_values))
+
+        violation_hard_constraints = np.concatenate(
+            (lower_violation_e[self.nonslacked_indices_e],
+             upper_violation_e[self.nonslacked_indices_e]))
+
+        # evaluate cost of soft constraints
+        lower_slack_cost_e, upper_slack_cost_e = np.array([0.]), np.array([0.])
+        if self.ocp.cost.Zl_e.size > 0:
+            lower_slack_cost_e += 0.5 * self.ocp.cost.Zl_e @ (lower_slack_e.full() * lower_slack_e.full())
+        if self.ocp.cost.zl_e.size > 0:
+            lower_slack_cost_e += self.ocp.cost.zl_e @ lower_slack_e.full()
+
+        if self.ocp.cost.Zu_e.size > 0:
+            upper_slack_cost_e += 0.5 * self.ocp.cost.Zu_e @ (upper_slack_e.full() * upper_slack_e.full())
+        if self.ocp.cost.zu_e.size > 0:
+            upper_slack_cost_e += self.ocp.cost.zu_e @ upper_slack_e.full()
+
+        if lower_slack_e.full().size > 0:
+            cost += lower_slack_e.full()
+        if upper_slack_e.full().size > 0 :
+            cost += upper_slack_e.full()
+
+        return cost[0][0]
+
+
+
+def get_path_cost_expression(ocp: AcadosOcp):
+    # multiply path costs with td, nonuniform grid
```

Comment:
I think these TODOs need to be addressed still :stuck_out_tongue_winking_eye:

Reply:
# multiply path costs with td, nonuniform grid -> this is done already
# multiply with cost scaling -> Implemented in an upcoming version ;)

Reply:
Then maybe add a list of _limitations_ to the pull request description, to document the next steps :)